src/rta/fullsupernodelist.cpp

```diff
@@ -0,0 +1,352 @@
+{
+    boost::shared_lock<boost::shared_mutex> readerLock(m_access);
+    auto it = m_list.find(address);
+    if (it != m_list.end())
+        return it->second;
+    return SupernodePtr(nullptr);
+}
+
+bool FullSupernodeList::buildAuthSample(uint64_t height, std::vector<FullSupernodeList::SupernodePtr> &out)
+{
+    crypto::hash block_hash;
+    std::string  block_hash_str;
+
+    if (!getBlockHash(height, block_hash_str)) {
+        LOG_ERROR("getBlockHash error");
+        return false;
+    }
+
+    epee::string_tools::hex_to_pod(block_hash_str, block_hash);
+    std::vector<SupernodePtr> tier_supernodes;
+
+
+    auto out_it = std::back_inserter(out);
+
+    auto build_tier_sample = [&](uint64_t tier_min, uint64_t tier_max) {
+        selectTierSupernodes(block_hash, tier_min, tier_max, tier_supernodes);
+        std::copy(tier_supernodes.begin(), tier_supernodes.end(), out_it);
+        tier_supernodes.clear();
+    };
+
+    build_tier_sample(Supernode::TIER1_STAKE_AMOUNT, Supernode::TIER2_STAKE_AMOUNT);
+    build_tier_sample(Supernode::TIER2_STAKE_AMOUNT, Supernode::TIER3_STAKE_AMOUNT);
+    build_tier_sample(Supernode::TIER3_STAKE_AMOUNT, Supernode::TIER4_STAKE_AMOUNT);
+    build_tier_sample(Supernode::TIER4_STAKE_AMOUNT, std::numeric_limits<uint64_t>::max());
+
+    return true;
+}
+
+std::vector<string> FullSupernodeList::items() const
+{
+    vector<string> result;
+    result.reserve(m_list.size());
+    boost::shared_lock<boost::shared_mutex> readerLock(m_access);
+    for (auto const& it: m_list)
+        result.push_back(it.first);
+
+    return result;
+}
+
+bool FullSupernodeList::getBlockHash(uint64_t height, std::string &hash)
+{
+    bool result = m_rpc_client.get_block_hash(height, hash);
+    return result;
+}
+
+std::future<void> FullSupernodeList::refreshAsync()
+{
+    m_refresh_counter = 0;
+    auto worker = [&](const std::string &address) {
+        SupernodePtr sn = this->get(address);
+        if (sn) {
+            sn->refresh();
+            ++m_refresh_counter;
+        }
+    };
+
+    for (const auto &address : this->items()) {
+        m_tp->enqueue(boost::bind<void>(worker, address));
+    }
+
+    return m_tp->runAsync();
+}
+
+size_t FullSupernodeList::refreshedItems() const
+{
+    return m_refresh_counter;
+}
+
+bool FullSupernodeList::bestSupernode(std::vector<SupernodePtr> &arg, const crypto::hash &block_hash, SupernodePtr &result)
+{
+    if (arg.size() == 0) {
+        LOG_ERROR("empty input");
+        return false;
+    }
+
+    std::vector<SupernodePtr>::iterator best = std::max_element(arg.begin(), arg.end(), [&](const SupernodePtr a, const SupernodePtr b) {
+        crypto::hash hash_a, hash_b;
+        a->getScoreHash(block_hash, hash_a);
+        b->getScoreHash(block_hash, hash_b);
+        uint256_t a_value = hash_to_int256(hash_a);
+        uint256_t b_value = hash_to_int256(hash_b);
+        // LOG_PRINT_L0("a_value: " << a_value << ", b_value: " << b_value);
```

Comment:
Are you still need it?

Reply:
removed with 0ace1d1098513e9c36a41244fe1f1a5ab16f13e8

Reply:
removed with 0ace1d1098513e9c36a41244fe1f1a5ab16f13e8